feat(Modal next): Introduce a next composable Modal (#9852)
patternfly/patternfly-react: packages/react-core/src/next/components/Modal/examples/Modal.md

---
id: Modal
section: components
cssPrefix: pf-v5-c-modal-box
propComponents: ['Modal', 'ModalBoxBody', 'ModalBoxHeader', 'ModalBoxFooter']
ouia: true
beta: true
---

import WarningTriangleIcon from '@patternfly/react-icons/dist/esm/icons/warning-triangle-icon';
import CaretDownIcon from '@patternfly/react-icons/dist/esm/icons/caret-down-icon';
import BullhornIcon from '@patternfly/react-icons/dist/esm/icons/bullhorn-icon';
import HelpIcon from '@patternfly/react-icons/dist/esm/icons/help-icon';
import spacing from '@patternfly/react-styles/css/utilities/Spacing/spacing';
import formStyles from '@patternfly/react-styles/css/components/Form/form';

## Examples

### Basic modals

> Basic modals give users the option to either confirm or cancel an action. To flag an open modal, use the `isOpen` property. To execute a callback when a modal is closed, use the `onClose` property.

Reviewer: ```suggestion
Basic modals give users the option to either confirm or cancel an action. 

To flag an open modal, use the `isOpen` property. To execute a callback when a modal is closed, use the `onClose` property.

A modal must have a `<ModalBoxBody>`, containing the main content of the modal. The `<ModalBoxHeader>` and `<ModalBoxFooter>` components are not required, but are typically used to display the modal title and any button actions, respectively. 
```
If we think this is a valuable callout to have here - vs if this is stated in the props table 🤷‍♀️

Author: I like this description.

```ts file="./ModalCustomTitleIcon.tsx"

```

### With wizard

To guide users through a series of steps in a modal, you can add a [wizard](/components/wizard) to a modal. To configure the `<Wizard>`, pass an array that contains a “name” and “component” value for each step into the `steps` property.

```ts file="./ModalWithWizard.tsx"

```

### With dropdown

To present a menu of actions or links to a user, you can add a [dropdown](/components/dropdown) to a modal. To allow the dropdown to visually break out of the modal container, set the `menuAppendTo` property to “parent”. Handle the modal’s closing behavior by listening to the `onEscapePress` callback on the `<Modal>` component. This allows the "escape" key to collapse the dropdown without closing the entire modal.

```ts file="./ModalWithDropdown.tsx"

```

### With help

To help simplify and explain complex models, add a help [popover](/components/popover). Only place a help icon at the modal level if its information applies to all content in the modal. If the help popover is specific to a particular modal section, place the help icon beside that section instead.

```ts file="./ModalWithHelp.tsx"

```

### With form

> To collect user input within a modal, you can add a [form](/components/form).

Reviewer: ```suggestion
To collect user input within a modal, you can add a [form](/components/forms/form).
```

> To submit the form from a button in the modal's footer (outside of the `<Form>`), set the button's `form` property equal to the form's id.

Reviewer: ```suggestion
To enable form submission from a button in the modal's footer (outside of the `<Form>`), set the button's `form` property equal to the form's id.
```

```ts file="ModalWithForm.tsx"

```

### Custom focus

> Use the `elementToFocus` property to customize which element inside the Modal receives focus when initially opened.

Reviewer: ```suggestion
To customize which element inside the modal receives focus when initially opened, use the `elementToFocus` property`. 
```

To adjust the size of a modal, use the `variant` property. Modal variants include "small", "medium", "large", and "default".

The following example displays a "small" modal by passing in `variant={ModalVariant.small}`.

```ts file="./ModalSmall.tsx"

```

### Medium modal

The following example displays a "medium" modal by passing in `variant={ModalVariant.medium}`.

```ts file="./ModalMedium.tsx"

```

### Large modal

The following example displays a "large" modal by passing in `variant={ModalVariant.large}`.

```ts file="./ModalLarge.tsx"

```

### Custom width

To choose a specific width for a modal, use the `width` property. The following example has a `width` of "50%".

```ts file="./ModalCustomWidth.tsx"

```

### Custom header and footer

To add a custom header and footer to a modal, set the `header` and `footer` properties to a custom implementation. The following example passes title components into both the header and the footer and also passes an icon to the footer.

```ts file="./ModalCustomHeaderFooter.tsx"

```

### No header or footer

To exclusively present information in a modal, remove the `header` and/or `footer`.

```ts file="./ModalNoHeaderFooter.tsx"

> ```

Author: I left both examples.  I Changed `Custom header and footer` to `Custom header` since there is nothing to pass to footer as a prop even when not custom.

I can remove the "No header or footer example if we think it adds no value.  Core however still has that exmple.

> ```

Reviewer: I think mentioning that ModalBoxHeader and ModalBoxFooter in the basic example sounds useful regardless. I'll add a suggestion abt that

It seems fine to keep/modify this example since it's calling out a specific situation/prop (title) that doesn't seem to be mentioned anywhere else already. Plus it parallels to the previous non-Next example?

But FYI the button in the example still says "header/footer", so it would need to be updated to just "header"!
